libcxx/include/__ranges/drop_view.h

```diff
@@ -90,6 +90,10 @@ namespace ranges {
       requires (!(__simple_view<_View> &&
                   random_access_range<const _View> && sized_range<const _View>))
     {
+      if constexpr (random_access_range<_View> && sized_range<_View>) {
+        range_difference_t<_View> __dist = ranges::distance(__base_);
```

Comment:
```suggestion
        const auto __dist = ranges::distance(__base_);
```

Reply:
I was like why don't you use `ranges::size` here and that is exactly what you want to call since that is what is constrained on.  then I realised that oh you want to call `std::min` on two signed types so you used `ranges::distance` just for converting the types.  maybe 
```range_difference_t<_View> __dist = ranges::size(__base_);```
is the actual intent of the code?

Reply:
> maybe
> `range_difference_t<_View> __dist = ranges::size(__base_);`
> is the actual intent of the code?

The standard does not guarantee that the above implicit conversion is well-formed. `ranges::distance` is always preferred over `ranges::size` when deal with iterators.

Reply:
> > maybe
> > `range_difference_t<_View> __dist = ranges::size(__base_);`
> > is the actual intent of the code?
> 
> The standard does not guarantee that the above implicit conversion is well-formed. `ranges::distance` is always preferred over `ranges::size` when deal with iterators.

Only narrowing conversion can be explicit-only.  IIUC if  size_t -> difeerence_t is narrowing conversion, I think we have other problems.  The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time. It is a utility tool to calculate the size in case you cannot caluculate it in constant time, which is the wrong tool here imo. And I strongly against using ‘auto’ when dealing with difference type for the exact reason we are discussing: narrowing conversion. 
If the spec does not say size -> difference_t cannot be narrowing, I’d prefer to use either parenthesis or static_cast to do the comversion

Reply:
> Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems. 

The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.

> The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time. 

This is why the `sized_range` needs to be satisfied in such case isn't it? in which case since `ranges::distance` is _guaranteed_ to be constant-time.

Reply:
> > Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems.
> 
> The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.
> 
> > The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time.
> 
> This is why the `sized_range` needs to be satisfied, isn't it? in which case `ranges::distance` is _guaranteed_ to be constant-time by its definition.

If you have a helper

Void helper() {
  if (c) 
       F();
   else
       G();
}

When you definitely don’t want to call G(), why do you write

if(c)
  helper()

Rather than call f() directly ?

Reply:
> > Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems.
> 
> The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.
> 
> > The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time.
> 
> This is why the `sized_range` needs to be satisfied, isn't it? in which case `ranges::distance` is _guaranteed_ to be constant-time by its definition.

Anyway I think our discussion about coding style should not block this patch to be landed

Reply:
> If you have a helper
> 
> Void helper() { if (c) F(); else G(); }
> 
> When you definitely don’t want to call G(), why do you write
> 
> if(c) helper()
> 
> Rather than call f() directly ?

This example is not quite accurate because the return type of `ranges::distance` is what we want, which reduces the type conversion since the former has already been done for us. Using `ranges::size` requires an additional `static_cast`.

I believe the standard prefers using `rangs::begin(r) + rangs::distance(r)` rather than `rangs::begin(r) + static_cast<range_difference_t<R>>(rangs::size(r))`, given the former's heavy presence in the `<ranges>`'s wording.

Reply:
I am fine with both. But I think the spec is describing the behaviour and not necessarily the way one should code. e.g this particular function here the spec is simply describing what the result should be, “Returns” , and one should not use the exact wording for implementation . (I saw you create an LWG issue for this but lots of us think it is not a defect because the spec does not say “equivalent to”)